### PR TITLE
Package magma: cleaned up cmake config

### PR DESCRIPTION
**Pull request by G-Ragghianti**
Implemented use of `define()` and `define_from_variant()` for cmake configuration, replaced `array.extend()` with `array.append()` where appropriate, and reduced some logical redundancy.

### CODE REVIEW

File: var/spack/repos/builtin/packages/magma/package.py [hunk 1]
```diff
@@ -106,55 +106,47 @@ def generate_gpu_config(self):
...
-                # explicitly add blas to LAPACK_LIBRARIES.
-                "-DLAPACK_LIBRARIES=%s" % (spec["lapack"].libs + spec["blas"].libs).joined(";"),
-            ]
-        )
-
-        options += ["-DBUILD_SHARED_LIBS=%s" % ("ON" if ("+shared" in spec) else "OFF")]
+        define = self.define
+        options = [
+            define("CMAKE_INSTALL_PREFIX", self.prefix),
+            define("CMAKE_INSTALL_NAME_DIR:PATH", self.prefix.lib),
```
Comment (haampie) [inline var/spack/repos/builtin/packages/magma/package.py]:
```suggestion
            define("CMAKE_INSTALL_NAME_DIR", self.prefix.lib),
```

File: var/spack/repos/builtin/packages/magma/package.py [hunk 2]
```diff
@@ -106,55 +106,47 @@ def generate_gpu_config(self):
...
+            define("BLAS_LIBRARIES", [spec["blas"].libs]),
```
Comment (haampie) [inline var/spack/repos/builtin/packages/magma/package.py]:
```suggestion
            define("BLAS_LIBRARIES", spec["blas"].libs),
```

File: var/spack/repos/builtin/packages/magma/package.py [hunk 3]
```diff
@@ -106,55 +106,47 @@ def generate_gpu_config(self):
...
+            # As of MAGMA v2.3.0, CMakeLists.txt does not use the variable
+            # BLAS_LIBRARIES, but only LAPACK_LIBRARIES, so we need to
+            # explicitly add blas to LAPACK_LIBRARIES.
+            define("LAPACK_LIBRARIES", [spec["lapack"].libs, spec["blas"].libs]),
```
Comment (haampie) [inline var/spack/repos/builtin/packages/magma/package.py]:
```suggestion
            define("LAPACK_LIBRARIES", spec["lapack"].libs + spec["blas"].libs),
```

File: var/spack/repos/builtin/packages/magma/package.py [hunk 4]
```diff
@@ -106,55 +106,47 @@ def generate_gpu_config(self):
...
+            options.append(define("USE_FORTRAN", True))
             if spec.satisfies("%xl") or spec.satisfies("%xl_r"):
-                options.extend(["-DCMAKE_Fortran_COMPILER=%s" % self.compiler.f77])
-
+                options.append(define("CMAKE_Fortran_COMPILER", self.compiler.f77))
             if spec.satisfies("%cce"):
-                options.append("-DCMAKE_Fortran_FLAGS=-ef")
+                options.append(define("CMAKE_Fortran_FLAGS", "-ef"))
 
         if spec.satisfies("^cuda"):
```
Comment (haampie) [inline var/spack/repos/builtin/packages/magma/package.py]: can you run this into `+cuda`, cause `^cuda` may match non-direct cuda deps

File: var/spack/repos/builtin/packages/magma/package.py [hunk 5]
```diff
@@ -106,55 +106,47 @@ def generate_gpu_config(self):
...
-            cuda_arch = self.spec.variants["cuda_arch"].value
-            if "@:2.2.0" in spec:
-                capabilities = " ".join("sm{0}".format(i) for i in cuda_arch)
-                options.extend(["-DGPU_TARGET=" + capabilities])
-            else:
-                capabilities = " ".join("sm_{0}".format(i) for i in cuda_arch)
-                options.extend(["-DGPU_TARGET=" + capabilities])
+            cuda_arch = spec.variants["cuda_arch"].value
+            sep = "" if "@:2.2.0" in spec else "_"
+            capabilities = " ".join("sm{0}{1}".format(sep, i) for i in cuda_arch)
```
Comment (haampie) [inline var/spack/repos/builtin/packages/magma/package.py:131]: should this be a list? i.e. semi-colon separated string?

**Review by haampie — CHANGES_REQUESTED** [review]
Thanks, some comments

File: var/spack/repos/builtin/packages/magma/package.py (hunk 5 above)
Comment (G-Ragghianti) [reply var/spack/repos/builtin/packages/magma/package.py:131]: The MAGMA CMakeLists.txt consistently uses spaces: https://bitbucket.org/icl/magma/src/master/CMakeLists.txt

File: var/spack/repos/builtin/packages/magma/package.py (hunk 4 above)
Comment (G-Ragghianti) [reply var/spack/repos/builtin/packages/magma/package.py]: I can do `"+cuda" in spec`, and I think that means roughly the same thing.  Is there a case where this isn't equivalent?

**Comment by G-Ragghianti** [comment]
> Thanks, some comments

Thank, submitted the changes.

**Comment by G-Ragghianti** [comment]
@haampie I assume this hasn't been merged because the state is still "requested changes", but I can't seem to get the state to change even though I've addressed all the suggested changes.

**Comment by bernhardkaindl** [comment]
@haampie: @G-Ragghianti says he adressed all suggested changes:
> @haampie I assume this hasn't been merged because the state is still "requested changes", but I can't seem to get the state to change even though I've addressed all the suggested changes.

**Review by luszczek — APPROVED** [review]

**Review by stomov — APPROVED** [review]

**Comment by G-Ragghianti** [comment]
Is this good?

**Comment by G-Ragghianti** [comment]
@bernhardkaindl Can this be merged?

**Review by bernhardkaindl — APPROVED** [review]
> @bernhardkaindl Can this be merged?

@haampie: Because you made a change request, you need to approve it was well, thanks!

**Review by haampie — APPROVED** [review]

**Pull request merged**